kafka-admin/src/main/java/org/bf2/admin/kafka/admin/RecordOperations.java

```diff
@@ -138,26 +139,41 @@ public String key() {
 
         CompletableFuture<Types.Record> promise = new CompletableFuture<>();
         Producer<String, String> producer = clientFactory.createProducer();
-        ProducerRecord<String, String> request = new ProducerRecord<>(topicName, input.getPartition(), stringToTimestamp(input.getTimestamp()), key, input.getValue(), headers);
-
-        producer.send(request, (meta, exception) -> {
-            if (exception != null) {
-                promise.completeExceptionally(exception);
-            } else {
-                Types.Record result = new Types.Record();
-                result.setPartition(meta.partition());
-                if (meta.hasOffset()) {
-                    result.setOffset(meta.offset());
-                }
-                if (meta.hasTimestamp()) {
-                    result.setTimestamp(timestampToString(meta.timestamp()));
-                }
-                result.setKey(input.getKey());
-                result.setValue(input.getValue());
-                result.setHeaders(input.getHeaders());
-                promise.complete(result);
+
+        try {
+            List<PartitionInfo> partitions = producer.partitionsFor(topicName);
+
+            if (partitions.isEmpty()) {
+                promise.completeExceptionally(noSuchTopic(topicName));
+            } else if (input.getPartition() != null && partitions.stream().noneMatch(p -> input.getPartition().equals(p.partition()))) {
+                promise.completeExceptionally(noSuchTopicPartition(topicName, input.getPartition()));
             }
-        });
+        } catch (TimeoutException e) {
+            promise.completeExceptionally(noSuchTopic(topicName));
+        }
+
+        if (!promise.isCompletedExceptionally()) {
```

Comment:
It feels slightly strange to be checking the state of the promise here and then conditionally continuing, but I'm not seeing a sensible refactoring to suggest.

Sorta feels like the try block is something like `validatePartition`  but that still doesn't chain usefully.

Reply:
Thanks for the feedback. I re-factored the actual send logic into a separate method to make this is a bit less icky...